nixos/doc: render option values using `lib.generators.toPretty`

ncfavier opened the pull request:
Fixes https://github.com/NixOS/nixpkgs/issues/195562, see discussion there.

A nice side effect is that we can leverage `toPretty`'s `allowPrettyValues` feature to make scrubbed derivations pretty-print as `pkgs.foo`. This should allow us to *mostly* get rid of the

```
default = pkgs.foo;
defaultText = lib.literalExpression "pkgs.foo";
```

pattern which is so easy to get wrong (though instances like `cfg.package` would still require it).

Note that if this approach works out, we can maybe turn it into a *promise* that `optionAttrSetToDocList` does not output non-rendered Nix values, so that we can eventually get rid of all the other Nix rendering code, in-tree and out-of-tree.

Regarding the first commit, I am aware that most of these issues would be avoided by adding proper type checks to defaults and examples. I'll try to look into this next.

ncfavier committed:
nixos/doc: fix some options

ncfavier committed:
nixos/doc: render option values using `lib.generators.toPretty`

Render un`_type`d defaults and examples as `literalExpression`s using
`lib.generators.toPretty` so that consumers don't have to reinvent Nix
pretty-printing. `renderOptionValue` is kept internal for now intentionally.

Make `toPretty` print floats as valid Nix values (without a tilde).

Get rid of the now-obsolete `substSpecial` function.

Move towards disallowing evaluation of packages in the manual by
raising a warning on `pkgs.foo.{outPath,drvPath}`; later, this should
throw an error. Instead, module authors should use `literalExpression`
and `mkPackageOption`.

ncfavier commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -286,6 +286,7 @@ in {
             log_config = mkOption {
               type = types.path;
               default = ./synapse-log_config.yaml;
+              defaultText = lib.literalExpression "nixos/modules/services/matrix/synapse-log_config.yaml";
```
I'm not sure what should be done with this. Maybe `toPretty` should somehow be made aware of `pkgs.path` so that it can render internal paths independently of where nixpkgs is located?

roberth replied:
The documentation generator already needs to deal with paths for the declaration and definition locations. The logic for this is a bit strewn about. It'd make sense to make the generic doc generation take ownership of that functionality and then reuse that logic by passing it to `toPretty`.

ncfavier committed:
lib/generators.toPretty: don't evaluate derivations

With the goal of making `toPretty` suitable for rendering option
values, render derivations as `<derivation foo-1.0>` instead of
`<derivation /nix/store/…-foo-1.0.drv>`.

This is to avoid causing sudden evaluation errors for out-of-tree
projects that have options with `default = pkgs.someUnfreePackage;` and
no `defaultText`.

pennae reviewed (approved):
looks reasonable, but we'd still rather not convert everything to text in nix. the generated manual does look a lot better on composite defaults though, so that's a big plus. might be worth it to make the average docs better, even if more advanced systems would have to re-parse examples.

> This should allow us to mostly get rid of the [defaultText for packages]

not really, since `pkgs` is a throw thunk in the lazy module docs build. might be better to use `mkPackageOption` instead for options referring to packages since that takes care of the duplication.

ncfavier commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -595,7 +595,8 @@ in
           type = types.package;
           default = cfg.host.pkgs.qemu_kvm;
-          example = "pkgs.qemu_test";
+          defaultText = literalExpression "config.virtualisation.host.pkgs.qemu_vm";
```
I think you meant `"config.${options.virtualisation.host.pkgs}.qemu_vm"`, but that fails with `attribute 'host' missing` for some reason.

ncfavier replied:
Yeah that doesn't work either because `virtualisation` is not an option.

The other error comes from the fact that we [override](https://github.com/NixOS/nixpkgs/blob/6a36558f5227f728d4fa433ab8ec1b80bebe4f50/nixos/modules/misc/documentation.nix#L43-L45) the `options` argument with the outer `options` so that eager modules have access to all options, but indeed "all" doesn't include the `qemu-vm.nix` module by default.

I looked into doing something like `options = lib.recursiveUpdate allOpts eagerOpts`, where `eagerOpts = scrubbedEval.options`, but this runs into infinite recursion because those are the *`checked`* options. We would need to be able to access the unchecked `options` fixed point, which could be done either by adding `uncheckedOptions = options;` [here](https://github.com/NixOS/nixpkgs/blob/6a36558f5227f728d4fa433ab8ec1b80bebe4f50/lib/modules.nix#L328-L333) or by allowing `specialArgs` to be a function that accepts `options`.

In any case this is way beyond the scope of this PR.

ncfavier commented:
> more advanced systems would have to re-parse examples.

What do you have in mind?

ncfavier commented on nixos/modules/misc/documentation.nix:
```diff
@@ -51,7 +51,11 @@ let
             if isAttrs value then
               scrubDerivations wholeName value
-              // (optionalAttrs (isDerivation value) { outPath = "\${${wholeName}}"; })
+              // optionalAttrs (isDerivation value) {
+                outPath = "\${${wholeName}}";
+                drvPath = wholeName;
+                __pretty = _: wholeName; val = null;
+              }
```
I think I agree. My main concern are things like `default = "... ${pkgs.foo} ..."` without a `defaultText`. The `"\${${wholeName}}"` hack currently makes this render "nicely" but this is arguably a bug with `toPretty`: the interpolation should be escaped, so that it doesn't look nice anymore. I think I'd like to do something like

```
// optionalAttrs (isDerivation value) {
  outPath = throw "attempt to use a derivation in options documentation";
  drvPath = # same
}
```

It's a bit annoying that this has to be decided at the level of the evaluation, rather than in the generic `make-options-doc`, so that out-of-tree projects have to make their own decision. But probably fine.

roberth replied:
> The `"\${${wholeName}}"` hack currently makes this render "nicely" but this is arguably a bug with `toPretty`:

I agree that this is a bug, but is it a reason to drop easy `pkgs.foo` interpolation from the docs?

It could be made to work well by
 - fixing the escaping first
 - making the escaping customizable (function passed to `toPretty`)
 - pass a custom escaping function, which at the end replaces a special syntax like `$$${{{` to `${` so that antiquotations are representable without stealing too much from the normal string literals.
 - make `pkgs.foo.outPath = "\$\$\${{{pkgs.foo}"`
 - toPretty will print the right expression

ncfavier replied:
Considering that this feature is not used by anything in-tree (manual still builds with a `throw`), I think it's worth trying to move in the direction of least hackiness, which IMO is something like https://github.com/NixOS/nix/issues/1474.

roberth replied:
Alright. Then let's keep using the least hacky option available with the Nix we have today:

```nix
{
  defaultText = lib.literalExpression ''
    "''${pkgs.foo}/bin/foo"
  '';
}
```

ncfavier committed:
lib/generators.toPretty: escape strings properly

ncfavier commented:
Alright, I got a bit carried away... see commits.

### Notes

- This PR changes the format of `options.json` (declarations can now be {path, url} instead of strings), which breaks nixos-search. This should be fixed *before* merging.
  Users of `optionAttrSetToDocList` should not get too broken (e.g. the home-manager manual still looks fine).
- After merging, the [release wiki](https://github.com/NixOS/release-wiki) should be updated to mention `lib.trivial.manualRevision`.
- After merging, morph can undo most of https://github.com/DBCDK/morph/pull/108/files and instead set `documentation.nixos.extraModuleSources = [ { name = "morph"; prefix = ../.; urlFor = m: ...; }]`. cc @bb010g
- https://github.com/NixOS/nix/pull/6530#issuecomment-1309963199 will break `lazy-options.json` because we now filter the nixpkgs source based on relative paths.
- An alternative to the `moduleSources` argument would be to add a `_module.sources` option so that modules can define their own sources. This could work either by also passing `config` to `optionsToDocTemplate`, or by making a `modulesToDocTemplate` wrapper which calls `evalModules`. I haven't thought this through, just throwing ideas around.

pennae reviewed (commented):
looks mostly reasonable. rendered options seem to fine judging by the limited number of checks we've done.

ncfavier commented on nixos/modules/misc/documentation.nix:
```diff
@@ -48,45 +47,46 @@ let
-          libPath = filter (pkgs.path + "/lib");
-          pkgsLibPath = filter (pkgs.path + "/pkgs/pkgs-lib");
-          nixosPath = filter (pkgs.path + "/nixos");
+          filteredNixpkgs = builtins.path {
+            name = "filtered-nixpkgs";
+            path = pkgs.path;
+            filter = let dirs = [ "lib/" "pkgs/pkgs-lib/" "nixos/" ]; in
+              path: type: let p = lib.removePrefix (toString pkgs.path + "/") (toString path); in
+              cleanSourceFilter path type && (
+                if type == "directory" then
+                  baseNameOf path != "tests" && lib.any (dir: lib.hasPrefix dir (p + "/") || lib.hasPrefix (p + "/") dir) dirs
+                else
+                  p == ".version" || lib.any (dir: lib.hasPrefix dir p) dirs);
+          };
```
The `.nix` filtering never worked, because the check was `t == "file"` instead of `t == "regular"`, and in fact excluding non-nix files isn't possible currently because of e.g. https://github.com/NixOS/nixpkgs/blob/cecc553095ea244e516003f7e0714669afb4af70/nixos/modules/services/system/nscd.nix#L57.

We could still make it work by adding a `defaultText` there, or marking that module non-cacheable, or we could just filter out directories named `doc`

ncfavier replied:
I think we should avoid bothering module authors with this implementation detail as much as possible, so I went with excluding `doc`. The resulting store path weighs 13.5 MiB, of which 12.6 MiB are .nix files.

ncfavier commented on nixos/modules/misc/documentation.nix:
```diff
@@ -294,16 +294,36 @@ in
+            };
+            name = mkOption {
+              type = with types; nullOr (strMatching "[a-zA-Z0-9._+-]+");
+              default = null;
```
The problem is compatibility: previously `extraModuleSources` only accepted paths to be stripped, so the question is what to do with those. Currently they are mapped to a nameless source, I don't see a better alternative short of throwing an error, but that seems a bit brutal.

bb010g replied:
`<_unknown/*>` may be a good default?

roberth replied:
Perhaps `baseNameOf config.prefix` and maybe strip off a store path hash part, if there is one.
I guess that will effectively be `"source"` in many cases.

What do the angle brackets, underscore, slash and asterisk mean? It seems a bit excessive.

bb010g replied:
@roberth Imagine a NixOS module source under version control, checked out in a home directory. If there's a module at the top level, e.g. `$CHECKOUT/configuration.nix`, the base name of `$CHECKOUT` isn't necessarily stable between checkouts. We should avoid looking at parents of the top-level. If we don't provide a default module source name, this path renders in documentation as `<configuration.nix>`. If the source then imports a module from `$CHECKOUT/nixpkgs/nixos/system/build.nix`, its path then renders as `<nixpkgs/nixos/system/build.nix>`, which is the path of an existing Nixpkgs module. (This example is contrived, but the situation can reasonably occur if multiple external module sets use simple paths.) On the other hand, if we provide a default module source name of `_unknown/`, then the former path renders as `<_unknown/configuration.nix>` and the latter path renders as `<_unknown/nixpkgs/nixos/system/build.nix>`.

This approach can still result in collisions, but a default name of `_unknown` sticks out and looks ugly enough that anyone who notices will change it, and the leading underscore is conventional for internal Nix & Nixpkgs names.

If we wanted the name to stick out more and be easier to search, `_unknown-module-source` could work really well (`<_unknown-module-source/configuration.nix>`).

On the `…/configuration.nix` note from earlier, we may want to add `config.nixos.extraModuleSources = [ { name = "nixos-config"; prefix = ./.; } ];` to the `/etc/nixos/configuration.nix` template.

roberth replied:
Ok, now I understand what you meant.

I'd be quite happy for it to throw an exception in projects where I control the scope of docs generation actually, but for NixOS, returning something ugly seems like a pretty good, robust solution.

We can't really guess the name, and we certainly can't let the NixOS evaluation fail because of the docs, so `<_unknown/whatever/path/file.nix>` seems like a good solution.

ncfavier replied:
Agreed, will do this.

> we may want to add `config.nixos.extraModuleSources = [ { name = "nixos-config"; prefix = ./.; } ];` to the `/etc/nixos/configuration.nix` template.

I'm not convinced. Seems like a lot of boilerplate for something the vast majority of people won't care about (and if that file is indeed destined to be in `/etc/nixos/configuration.nix`, then `<nixos-config/X>` isn't a lot better than `/etc/nixos/X`.

ncfavier replied:
Actually, here's an alternative: introduce a new `moduleSources` option with the "ideal" type (attrsOf {prefix, urlFor}), and add a `mkChangedOptionModule` that maps paths from `extraModuleSources` to attributes in `moduleSources` using the mechanism proposed by @roberth (basename of prefix minus store hash). Seems like the best of all worlds?

ncfavier replied:
I guess this will break if someone has two paths named `source`, which is quite likely...

roberth replied:
I thought `name` was only a label for presentation, not a lookup key. Isn't prefix the only lookup key?

roberth replied:
Lookup key in the logical sense. I think prefix has to be a linear search, not an attrset key, because it can be a path value, and we may not want to `toString` those because of lazy trees.

ncfavier replied:
See the mirror [discussion](https://github.com/NixOS/nixpkgs/pull/199363#discussion_r1021733671) for `searchPath`. Technically you're right, but I think it makes sense to have module sources indexed by their name. We might find more use cases for them later than just stripping prefixes.

roberth replied:
You could use `name` as an extra key that's an actual key and have a `displayName` option for the cases where we don't have enough info.

ncfavier replied:
Not sure I understand the use case for that, and in any case it seems well beyond the scope of this PR so I think I'll revert to the "least disruptive" choice (the current state of this PR, but make `name` non-optional and default to `_unknown` for raw paths) and let others extend this later.

ncfavier commented on lib/generators.nix:
```diff
@@ -278,40 +278,54 @@ rec {
+         toPretty { searchPath = [ { name = "foo"; prefix = "/path/to/foo"; } ]; } /path/to/foo/bar/qux
+         ==> "<foo/bar/qux>"
+       Type: [ { name = string; prefix = string or path; } ] */
+    searchPath ? []
```
I guess *semantically* this should be a mapping from names to paths, but *technically* it should be a mapping from paths to names (because that's the direction of the transformation we're doing), so I went the neutral route and picked neither.

I'm fine with going the semantic route though.

ncfavier commented:
I went with my `_module.sources` idea and isolated the prefix lookup logic into a new function `lib.strings.lookupPrefix`.
That way it's not NixOS-specific, module sources are responsible for identifying themselves, and we can bake the nixpkgs source in (which will always be used for the `_module` options if those are rendered).

Also moved the `<source/path>` rendering down to XSLT; other consumers get a {source, path, url?} attrset which they are free to render as they wish.

I still think it's fine to allow sources with colliding `name`s, e.g. when Nix gets lazy trees we might somehow want to have roots for both `"${flake}"` and `toString flake` in certain situations.

I've tested this in a variety of situations and I've got PRs ready for home-manager, nmd and nixos-search.

fricklerhandwerk commented:
Oof, that's a large one. I suppose I was added automatically based on `CODEOWNERS`. Sorry @ncfavier, I won't have time to review this any time soon, maybe never, but you seem to have plenty of reviewers who can help merge this.

infinisil commented:
This PR does too many things now! Can you split this up into smaller reviewable bits @ncfavier?

ncfavier commented:
Indeed, forked the path stuff into https://github.com/NixOS/nixpkgs/pull/203994.

Note that the commits left in this PR are already reviewed (but additional reviews can't hurt!)

ncfavier committed:
nixos/make-options-doc: remove dead code

The logic for pretty-printing Nix values isn't needed any more because
`optionAttrSetToDocList` returns already rendered values.

ncfavier commented on nixos/lib/make-options-doc/options-to-docbook.xsl:
```diff
@@ -138,82 +138,6 @@
-  </xsl:template>
-
-
-  <xsl:template match="attrs[attr[@name = '_type' and string[@value = 'literalExpression']]]">
```
Note there's already a template for `literalExpression` above.

ncfavier commented:
I think this is good to merge now?

pennae reviewed (commented):
yep, looks good (except for a few nitpicks. just ignore them if you want to get this over with.)

ncfavier commented on nixos/modules/services/databases/openldap.nix:
```diff
@@ -16,7 +16,7 @@ let
       # systemd/systemd#19604
       description = ''
         LDAP value - either a string, or an attrset containing
-        <literal>path</literal> or <literal>base64</literal> for included
```
I know, but raw MarkDown is at least more readable than raw DocBook.

ncfavier commented on lib/generators.nix:
```diff
@@ -294,7 +297,7 @@ rec {
             outroSpace = if multiline then "\n${indent}" else " ";
     in if   isInt      v then toString v
-    else if isFloat    v then "~${toString v}"
+    else if isFloat    v then toString v
```
Alright, I went with `toJSON` and left a comment
